plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,464 @@
+{
+	g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
+	g_autoptr(GFile) mbim_device_file = g_file_new_for_path (self->mbim_port);
+	OpenContext ctx = {
+		.mainloop = mainloop,
+		.mbim_device = NULL,
+		.error = NULL,
+		.open_attempts = FU_MBIM_QDU_MAX_OPEN_ATTEMPTS,
+	};
+
+	mbim_device_new (mbim_device_file, NULL, fu_mbim_qdu_updater_mbim_device_new_ready, &ctx);
+	g_main_loop_run (mainloop);
+
+	/* either we have all device or otherwise error is set */
+	if (ctx.mbim_device != NULL) {
+		g_warn_if_fail (!ctx.error);
+		self->mbim_device = ctx.mbim_device;
+		/* success */
+		return TRUE;
+	}
+
+	g_warn_if_fail (ctx.error != NULL);
+	g_warn_if_fail (ctx.mbim_device == NULL);
+	g_propagate_error (error, ctx.error);
+	return FALSE;
+}
+
+typedef struct {
+	GMainLoop	*mainloop;
+	MbimDevice	*mbim_device;
+	guint		 timeout_id;
+	GError		*error;
+} CloseContext;
+
+static void
+fu_mbim_qdu_updater_mbim_device_close_ready (GObject *mbim_device, GAsyncResult *res, gpointer user_data)
+{
+	CloseContext *ctx = (CloseContext *) user_data;
+
+	/* ignore errors when closing */
+	mbim_device_close_finish (MBIM_DEVICE (mbim_device), res, (ctx->error == NULL) ? &ctx->error : NULL);
+	g_clear_object (&ctx->mbim_device);
+	g_main_loop_quit (ctx->mainloop);
+}
+
+static gboolean
+fu_mbim_qdu_updater_reprobe_timeout (gpointer user_data)
+{
+	CloseContext *ctx = user_data;
+
+	ctx->timeout_id = 0;
+	g_main_loop_quit (ctx->mainloop);
+
+	return G_SOURCE_REMOVE;
+}
+
+gboolean
+fu_mbim_qdu_updater_close (FuMbimQduUpdater *self, GError **error)
+{
+	g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
+	CloseContext ctx = {
+		.mainloop = mainloop,
+		.mbim_device = g_steal_pointer (&self->mbim_device),
+		.timeout_id = 0,
+	};
+
+	mbim_device_close (ctx.mbim_device, 5, NULL, fu_mbim_qdu_updater_mbim_device_close_ready, &ctx);
+	g_main_loop_run (mainloop);
+
+	/* we should always have both device cleared, and optionally error set */
+	g_warn_if_fail (ctx.mbim_device == NULL);
+
+	if (ctx.error != NULL) {
+		g_propagate_error (error, ctx.error);
+		return FALSE;
+	}
+
+	g_warn_if_fail (ctx.timeout_id == 0);
+	ctx.timeout_id = g_timeout_add_seconds (15, fu_mbim_qdu_updater_reprobe_timeout, &ctx);
+	g_main_loop_run (mainloop);
+
+	/* update detach right after this */
+	return TRUE;
+}
+
+typedef struct {
+	GMainLoop	*mainloop;
+	MbimDevice	*mbim_device;
+  	GCancellable    *cancellable;
+	GError		*error;
+	guint		 timeout_id;
+	GBytes		*blob;
+	GArray		*digest;
+	GPtrArray	*chunks;
+	guint 		chunk_sent;
+	FuDevice 	*device;
+} WriteContext;
+
+static gboolean
+fu_mbim_qdu_updater_reboot_timeout (gpointer user_data)
+{
+	WriteContext *ctx = user_data;
+
+	ctx->timeout_id = 0;
+	g_main_loop_quit (ctx->mainloop);
+
+	return G_SOURCE_REMOVE;
+}
+
+static void
+fu_mbim_qdu_updater_file_write_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	g_autoptr(MbimMessage) request  = NULL;
```

Comment:
If I'm not mistaken, in fwupd all the g_autoptr() variables are grouped together at the end of the list of variables.

Reply:
Also, some of these variable names are aligned and some others not, maybe do the same thing for all?